drone/server.go

```diff
@@ -300,19 +313,79 @@ func server(c *cli.Context) error {
 		middleware.Broker(c),
 	)
 
-	// start the server with tls enabled
-	if c.String("server-cert") != "" {
-		return http.ListenAndServeTLS(
-			c.String("server-addr"),
-			c.String("server-cert"),
-			c.String("server-key"),
-			handler,
+	if c.Bool("lets-encrypt-enabled") || (c.String("server-cert") != "" && c.String("server-key") != "") {
+		// define proper accepted curves
+		curves := []tls.CurveID{
+			tls.CurveP521,
+			tls.CurveP384,
+			tls.CurveP256,
+		}
+
+		// define proper accepted ciphers
+		ciphers := []uint16{
+			tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+			tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+		}
+
+		cfg := &tls.Config{
+			PreferServerCipherSuites: true,
+			MinVersion:               tls.VersionTLS12,
+			CurvePreferences:         curves,
+			CipherSuites:             ciphers,
+		}
+
+		if c.Bool("lets-encrypt-enabled") {
+			if c.String("lets-encrypt-path") == "" {
+				return fmt.Errorf("No Let's Encrypt cert storage path defined")
+			}
+
+			certManager := autocert.Manager{
+				Prompt: autocert.AcceptTOS,
+				Cache:  autocert.DirCache(c.String("lets-encrypt-path")),
+			}
+
+			cfg.GetCertificate = certManager.GetCertificate
+		} else {
+			cert, err := tls.LoadX509KeyPair(
+				c.String("server-cert"),
+				c.String("server-key"),
+			)
+
+			if err != nil {
+				return fmt.Errorf("Failed to load SSL certificates. %s", err)
+			}
+
+			cfg.Certificates = []tls.Certificate{
+				cert,
+			}
+		}
+
+		// define the server configuration
+		server := &http.Server{
+			Addr:         c.String("server-addr"),
```

Comment:
I'm wondering, with let's encrypt, do we want to use `https:` as the address, and then start up a second server that listens for `http:` and auto-redirects to https?